File: NeetCode 150/Graphs/743. Network Delay Time.py
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        #Solution is BFS + Min Heap
        adj = defaultdict(list)
        for u,v,w in times:
            adj[u].append((v,w))
        
        min_heap = [(0,k)] #We are not using queue in BFS, we are using min_heap
        time = 0
        visit = set()
        while min_heap:
            w1,n1 = heapq.heappop(min_heap)
            if n1 in visit:
                continue
            visit.add(n1)
            time = max(time,w1)
            #in Heap we have w,n and in adj list we have n,w
            for n2,w2 in adj[n1]:
                if n2 not in visit:
                    heapq.heappush(min_heap,(w1+w2,n2))
                    
        return time if len(visit) == n else -1

File: NeedCode 150 (JAVA)/Graphs/743. Network Delay Time.java
/*
Why Dijkstra?

Because all edge weights are non-negative.

BFS does not work for weighted edges.
Dijkstra finds the shortest distance from one source to all nodes.
*/
class Solution {
    public int networkDelayTime(int[][] times, int n, int k) {

    List<List<int[]>> graph = new ArrayList<>();
    for(int i =0;i<=n;i++){//nodes are labeled from 1 to n. we dont use 0, but 1 to n we use
        graph.add(new ArrayList<>());  
    }

    for(int[] time: times){
        int u = time[0];
        int v = time[1];
        int w = time[2];

        graph.get(u).add(new int[]{v,w});
    }

    PriorityQueue<int[]> pq = new PriorityQueue<>((a,b) -> a[0] - b[0]);
    pq.offer(new int[]{0,k});

    HashSet<Integer> visited = new HashSet<>(); //you can use boolean visited, but then you will also need count to check at the end
    int time = 0;

    while(!pq.isEmpty() && visited.size() < n){
        int[] curr = pq.poll();
        int d1 = curr[0];
        int node = curr[1];

        if(visited.contains(node))
            continue;
        visited.add(node);
        time = Math.max(time,d1); //the new node we are currently visting might have more time. so max. if We are polling any node. that is anyway the least time until now to reach that node.

        for(int[] nei: graph.get(node)){
            int next = nei[0];
            int d2 = nei[1];
            if(!visited.contains(next)){
                pq.offer(new int[]{d1+d2,next});
            }
        }
    }

    return visited.size() == n  ? time : -1;

    }
}
